File: lista_3/zadanie_1/java/src/FiniteField.java
public class FiniteField implements Comparable<FiniteField> {
    private final long p;
    private final long value;

    private static long betterMod(long a, long p) {
        long result = a % p;
        if (result < 0) {
            result += p;
        }
        return result;
    }

    private static long betterPow(long a, long b, long p) {
        long result = 1;
        while (b > 0) {
            if (b % 2 == 1) {
                result = betterMod(result * a, p);
            }
            a = betterMod(a * a, p);
            b /= 2;
        }
        return result;
    }

    public FiniteField(long value, long prime) {
        if (!isPrime(prime)){
            throw new IllegalArgumentException("The number is not prime");
        }
        this.p = prime;
        this.value = betterMod(value, p);
    }

    private static boolean isPrime(long p) {
        if (p == 2) {
            return true;
        }
        if (p % 2 == 0) {
            return false;
        }
        for (long i = 3; i * i <= p; i += 2) {
            if (p % i == 0) {
                return false;
            }
        }
        return true;
    }

    public long getValue() {
        return value;
    }

    public FiniteField add(FiniteField other) {
        if (p != other.p) {
            throw new IllegalArgumentException("The numbers are not from the same field");
        }
        return new FiniteField(this.value + other.getValue(), p);
    }

    public FiniteField subtract(FiniteField other) {
        if (p != other.p) {
            throw new IllegalArgumentException("The numbers are not from the same field");
        }
        return new FiniteField(this.value - other.getValue(), p);
    }

    public FiniteField multiply(FiniteField other) {
        if (p != other.p) {
            throw new IllegalArgumentException("The numbers are not from the same field");
        }
        return new FiniteField(this.value * other.getValue(), p);
    }

    public FiniteField inverse() {
        if (this.value == 0) {
            throw new ArithmeticException("Division by zero");
        }
        return new FiniteField(betterPow(this.value, p - 2, p), p);
    }

    public FiniteField divide(FiniteField other) throws ArithmeticException {
        if (p != other.p) {
            throw new IllegalArgumentException("The numbers are not from the same field");
        }
        if (other.getValue() == 0) {
            throw new ArithmeticException("Division by zero");
        }
        return this.multiply(other.inverse());
    }

    public FiniteField pow(int exponent) {
        return new FiniteField(betterPow(this.value, exponent, p), p);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof FiniteField other) {
            if (p != other.p) {
                return false;
            }
            return this.value == other.getValue();
        }

        return super.equals(obj);
    }

    @Override
    public String toString() {
        return String.valueOf(value);
    }

    @Override
    public int compareTo(FiniteField o) {
        if (p != o.p) {
            throw new IllegalArgumentException("The numbers are not from the same field");
        }
        return Long.compare(this.value, o.getValue());
    }

}

File: lista_3/zadanie_1/c++/finite_field.cpp
#include "finite_field.h"

bool finite_field::isPrime(long number)
{
    if (number <= 1)
        return false;
    if (number <= 3)
        return true;
    if (number % 2 == 0 || number % 3 == 0)
        return false;
    for (long i = 5; i * i <= number; i = i + 6)
        if (number % i == 0 || number % (i + 2) == 0)
            return false;
    return true;
}

finite_field::finite_field(long value, long p) : p(p), value(value % p)
{
    if (!isPrime(p))
        throw std::invalid_argument("p is not prime");
}

finite_field::~finite_field()
{
}

unsigned long finite_field::getValue() const
{
    return value;
}

unsigned long finite_field::getP() const
{
    return p;
}

finite_field finite_field::pow(finite_field exponent)
{
    if (exponent.getP() != p)
    {
        throw std::invalid_argument("Exponent must be in the same field");
    }
    unsigned long result = 1;
    unsigned long base = value;
    unsigned long exp = exponent.getValue();
    while (exp > 0)
    {
        if (exp % 2 == 1)
            result = (result * base) % p;
        exp = exp >> 1;
        base = (base * base) % p;
    }
    return finite_field(result, p);
}

finite_field finite_field::inverse()
{
    if (value == 0)
        throw std::invalid_argument("Cannot calculate inverse of 0");
    return pow(finite_field(p - 2, p));
}

finite_field finite_field::operator+(finite_field other)
{
    if (other.getP() != p)
    {
        throw std::invalid_argument("Cannot add elements from different fields");
    }
    return finite_field(value + other.getValue() % p, p);
}

finite_field finite_field::operator-(finite_field other)
{
    if (other.getP() != p)
    {
        throw std::invalid_argument("Cannot subtract elements from different fields");
    }
    return finite_field(value - other.getValue() % p, p);
}

finite_field finite_field::operator*(finite_field other)
{
    if (other.getP() != p)
    {
        throw std::invalid_argument("Cannot multiply elements from different fields");
    }
    return finite_field(value * other.getValue() % p, p);
}

finite_field finite_field::operator/(finite_field other)
{
    if (other.getP() != p)
    {
        throw std::invalid_argument("Cannot divide elements from different fields");
    }
    if (other.getValue() == 0)
        throw std::invalid_argument("Division by zero");
    return finite_field(value * other.inverse().getValue() % p, p);
}

bool finite_field::operator==(finite_field other)
{
    if (other.getP() != p)
    {
        throw std::invalid_argument("Cannot compare elements from different fields");
    }
    return value == other.getValue();
}

bool finite_field::operator!=(finite_field other)
{
    if (other.getP() != p)
    {
        throw std::invalid_argument("Cannot compare elements from different fields");
    }
    return value != other.getValue();
}

bool finite_field::operator<(finite_field other)
{
    if (other.getP() != p)
    {
        throw std::invalid_argument("Cannot compare elements from different fields");
    }
    return value < other.getValue();
}

bool finite_field::operator>(finite_field other)
{
    if (other.getP() != p)
    {
        throw std::invalid_argument("Cannot compare elements from different fields");
    }
    return value > other.getValue();
}

bool finite_field::operator<=(finite_field other)
{
    if (other.getP() != p)
    {
        throw std::invalid_argument("Cannot compare elements from different fields");
    }
    return value <= other.getValue();
}

bool finite_field::operator>=(finite_field other)
{
    if (other.getP() != p)
    {
        throw std::invalid_argument("Cannot compare elements from different fields");
    }
    return value >= other.getValue();
}

std::ostream &operator<<(std::ostream &os, const finite_field &field)
{
    os << field.getValue();
    return os;
}
